api/dataset.go

```diff
@@ -1068,10 +1074,66 @@ func (d *PublishCheck) Check(handle func(http.ResponseWriter, *http.Request)) ht
 			return
 		}
 
-		if versionDoc != nil {
-			if versionDoc.State == models.PublishedState {
+		if currentVersion != nil {
+			if currentVersion.State == models.PublishedState {
+				defer func() {
+					if err := r.Body.Close(); err != nil {
+						log.ErrorC("could not close response body", err, nil)
+					}
+				}()
+
+				versionDoc, err := models.CreateVersion(r.Body)
+				if err != nil {
+					log.ErrorC("failed to model version resource based on request", err, log.Data{"dataset_id": id, "edition": edition, "version": version})
+					http.Error(w, err.Error(), http.StatusBadRequest)
+					return
+				}
+
+				// We can allow public download links to be modified by the exporters when a version is published.
+				// Note that a new version will be created which contain only the download information to prevent
+				// any forbidden fields from being set on the published version
+				if versionDoc.Downloads != nil {
+					newVersion := new(models.Version)
+					if versionDoc.Downloads.CSV != nil && versionDoc.Downloads.CSV.Public != "" {
+						newVersion = &models.Version{
+							Downloads: &models.DownloadList{
+								CSV: &models.DownloadObject{
+									Public: versionDoc.Downloads.CSV.Public,
+									Size:   versionDoc.Downloads.CSV.Size,
+									HRef:   versionDoc.Downloads.CSV.HRef,
+								},
+							},
+						}
+					}
+					if versionDoc.Downloads.XLS != nil && versionDoc.Downloads.XLS.Public != "" {
+						newVersion = &models.Version{
+							Downloads: &models.DownloadList{
+								XLS: &models.DownloadObject{
+									Public: versionDoc.Downloads.CSV.Public,
+									Size:   versionDoc.Downloads.CSV.Size,
+									HRef:   versionDoc.Downloads.CSV.HRef,
+								},
+							},
+						}
+					}
+					if newVersion != nil {
+						b, err := json.Marshal(newVersion)
+						if err != nil {
+							http.Error(w, err.Error(), http.StatusForbidden)
+							return
+						}
+
+						if err := r.Body.Close(); err != nil {
+							log.ErrorC("could not close response body", err, nil)
+						}
+						r.Body = ioutil.NopCloser(bytes.NewBuffer(b))
+						handle(w, r)
+						return
+					}
+				}
+
 				err = errors.New("unable to update version as it has been published")
-				log.Error(err, log.Data{"version": versionDoc})
+				log.Error(err, log.Data{"version": currentVersion})
```

Comment:
It might be worth logging both current version and the versionDoc here to see what the request was trying to update?